### ISSUE TITLE
When creating an annotated tag, a TAGMESSAGE file is left in the repository

### ISSUE THREAD

**Issue by forbjok** [gitextensions/gitextensions#4358]
When creating an annotated tag, a TAGMESSAGE file is left in the repository afterwards.

- GitExtensions version: 2.51RC1
- GIT version: 2.15.1.2
- OS version: Windows 10 Pro version 1709

**Did this work in previous version of GitExtensions (which)?**
This did not happen in any previous version of GitExtensions that I know of. (such as 2.50.2)

**Comment by glen-nicol** [comment]
I would like to work on this. It seems the actual change that caused this is the fact that the TAGMESSAGE file is no longer placed in the .git directory but in the repo itself. 9dc2fe18b1ce56e09a7b30f2f05a86d58956e602. Specifically the change from [workingDirGitDir to workingdir](https://github.com/gitextensions/gitextensions/commit/9dc2fe18b1ce56e09a7b30f2f05a86d58956e602#diff-0e78113ee21f8850bd28baab3531790fR40).  The tests still indicate the file should be [in a .git directory](https://github.com/gitextensions/gitextensions/blob/master/UnitTests/GitCommandsTests/Git/Tag/GitCreateTagCmdTests.cs#L14). So that may need to change based on the approach we want to take for solving this.
1. Put it back into the .git directory? 
2. Add handling to the command or controller to be able to clean that file up. I couldn't find anything that ever did that.
3.  Rename the file to TAGMESSAGE.tmp which most gitignores include. Kind of a lame solution though.

**Comment by gerhardol** [comment]
The temporary file should probably be requested as a separate service like System.IO.Path.GetTempPath() so it does not matter so much if a problem occurs. The location of the file can possibly be done in such way that the path is not included in the testcase.

The information that the file is temporary should be preserved so the temp file can be deleted after.

**Comment by glen-nicol** [comment]
So what are you proposing exactly? How would this service know that it is okay to delete the file?

Seems like we need to have a function on the controller that allows us to notify it that the tag command has completed. Then it can notify the service the file is no longer needed.  Would need to make that call in [FormCreateTag.cs](https://github.com/gitextensions/gitextensions/blob/a2020849041da641e92dc6278598db71af9ad449/GitUI/CommandsDialogs/FormCreateTag.cs#L88) and maybe [in FormVerify.cs](https://github.com/gitextensions/gitextensions/blob/a2020849041da641e92dc6278598db71af9ad449/GitUI/CommandsDialogs/FormCreateTag.cs#L88). 

Another way I see we could do it is to make [GitTagCommand](https://github.com/gitextensions/gitextensions/blob/a2020849041da641e92dc6278598db71af9ad449/GitCommands/Git/Tag/GitCreateTagCmd.cs#L6) IDisposable so that the callers need to dispose of it when they are done and that file can be deleted that way. But that requires either the new service to return an Idisposable object that GitCreateTagCmd can dispose of or we have to put a reference to the new service in GitCreateTagCmd.

**Comment by gerhardol** [comment]
I have not thought this through, but when FormCreateTag has executed the command, it should call a new method in TagController to cleanup (or use IDisposable). Or provide the temporary file to TagController (the temp file can be ignored when out of scope).

**Comment by RussKie** [comment]
I'll have a look and share my thoughts

**Comment by RussKie** [comment]
Sorry Glen, I haven't forgotten about you.
I'm flat out with work atm doing 12 hours shifts....

Feel free to propose a solution, else I'll try to help you as soon as my
work quiets down

**Comment by glen-nicol** [comment]
Could we move all the logic into the controller? So the ui just calls into the controller where it can keep track of the file and can delete it if necessary. That way the controller doesn't have to have a callback with info and we don't have to make a disposable type. 

I forget what the interface to launch the command is. If that isn't a static it could be passed into the controller along with the ui form arguments.

**Comment by RussKie** [comment]
How do you propose we track it at the controller level?

I had a quick looks at the code, and the idea of making the `GitCreateTagCmd : IDisposable` does not sound crazy at all...
![image](https://user-images.githubusercontent.com/4403806/36594854-02ec84b8-18f4-11e8-8f2d-a4312b78cdc2.png)
The command is used in two different places (ignore the unit test), and it feels sensible to unify the disposal behaviour. The way it is used seems to ask for `IDisposable` implementations too.
My 2c, happy to be challenged or corrected.

**Comment by glen-nicol** [comment]
Take a look at [my fork](https://github.com/glen-nicol/gitextensions/commit/c790a16db17b72290ac9d7b7ba8d3a59c5c01ffe).

**Comment by RussKie** [comment]
Please raise a PR, then everyone can look at it and comment.

**Comment by glen-nicol** [comment]
Done #4526

**Comment by jan-r** [comment]
Would it be possible to get this fix in the next 2.51 maintenance release?

**Comment by jan-r** [comment]
Great, thanks!

**Comment by RussKie** [comment]
@jan-r could you please test this build and confirm the issue is fixed
https://ci.appveyor.com/api/buildjobs/hqxlol9n99ka5g3b/artifacts/GitExtensions-2.51.01.2331-Mono.zip

**Comment by jan-r** [comment]
@RussKie tested on Linux and Win 10, works like a charm! Thanks again for the quick response!

**Comment by RussKie** [comment]
Awesome, thanks
We'll have it released in the next maintenance release

**Comment by glen-nicol** [comment]
So this is slightly embarrassing. At work a coworker and I are still getting TAGMESSAGE files left around after updating to 2.51.01. Works just fine at home where I did my patch. Home computer is Windows 10. Work is Windows 7. Coworker wondered if it could be a race condition with the file not being created when the _filesystem.Delete call goes through. But I don't see how that could be if the other window spawns and reads from the file to make the tag. And if the file was in use then we would get an exception but we don't get any popups.

The path is stored in a local variable that is never changed so our working dir can't change out from underneath us like it could if we recalled:

```csharp
private string GetWorkingDirPath()
{
    return _uiCommands.GitModule.WorkingDir;
}
```

Also worth noting that at least in the repo I am using at work there are a few submodules.

Any ideas on what else could cause that finally{ delete } to fail or where to look for signs of failure?
